From 6d3838ec2555e0294d90f3a5ef4487f6d1b00138 Mon Sep 17 00:00:00 2001
From: "Eric S. Raymond" <esr@thyrsus.com>
Date: Sat, 26 Nov 2016 00:29:50 -0500
Subject: [PATCH 014/268] Repair mrulist reporting.

The ntp_monitor() call got pushed down a code path that packets (at least,
packets from pool servers) don't normally take.  Result: mosdt traffic
was not getting recorded, and ntpq -c mrulist returned spuriously empty
reports.

Fix verified by actually seeing an mrulist cpme out afterwards.
---
 ntpd/ntp_proto.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/ntpd/ntp_proto.c b/ntpd/ntp_proto.c
index c5e37fe..ae7d56c 100644
--- a/ntpd/ntp_proto.c
+++ b/ntpd/ntp_proto.c
@@ -477,12 +477,6 @@ handle_fastxmit(
 		}
 	}
 
-	restrict_mask = ntp_monitor(rbufp, restrict_mask);
-	if (restrict_mask & RES_LIMITED) {
-		sys_limitrejected++;
-		if(!(restrict_mask & RES_KOD)) { return; }
-	}
-
 	/* To prevent exposing an authentication oracle, only MAC
 	   the response if the request passed authentication.
 	*/
@@ -793,6 +787,12 @@ receive(
 		}
 	}
 
+	restrict_mask = ntp_monitor(rbufp, restrict_mask);
+	if (restrict_mask & RES_LIMITED) {
+		sys_limitrejected++;
+		if(!(restrict_mask & RES_KOD)) { goto done; }
+	}
+
 	switch(match) {
 	    case AM_FXMIT:
             case AM_NEWPASS:
-- 
2.7.4

